google/cloud/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,275 @@
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_RESUMABLE_ASYNC_STREAMING_READ_WRITE_RPC_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_RESUMABLE_ASYNC_STREAMING_READ_WRITE_RPC_H
+
+#include "google/cloud/async_streaming_read_write_rpc.h"
+#include "google/cloud/backoff_policy.h"
+#include "google/cloud/internal/async_sleeper.h"
+#include "google/cloud/internal/retry_policy.h"
+#include "google/cloud/status_or.h"
+#include "google/cloud/version.h"
+#include <chrono>
+#include <memory>
+#include <thread>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+/**
+ * `ResumableAsyncStreamingReadWriteRpc<ResponseType, RequestType>` uses
+ * callables compatible with this `std::function<>` to create new streams.
+ */
+template <typename ResponseType, typename RequestType>
+using AsyncStreamFactory = std::function<
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>()>;
+
+/**
+ * `ResumableAsyncStreamingReadWriteRpc<ResponseType, RequestType>` uses
+ * callables compatible with this `std::function<>` to update the request object
+ * after each response. This is how users of the class can update the resume
+ * token or any other parameters needed to restart a stream from the last
+ * received message.
+ */
+template <typename ResponseType, typename RequestType>
+using StreamReinitializer = std::function<future<StatusOr<
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>>>(
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>)>;
+
+template <typename ResponseType, typename RequestType>
+class ResumableAsyncStreamingReadWriteRpc
+    : public AsyncStreamingReadWriteRpc<RequestType, ResponseType> {
+ public:
+  virtual future<Status> Status() = 0;
+};
+
+template <typename ResponseType, typename RequestType>
+class ResumableAsyncStreamingReadWriteRpcImpl
+    : public ResumableAsyncStreamingReadWriteRpc<RequestType, ResponseType> {
+ public:
+  ResumableAsyncStreamingReadWriteRpcImpl(
+      std::unique_ptr<RetryPolicy const> retry_policy,
+      std::unique_ptr<BackoffPolicy const> backoff_policy,
+      const AsyncSleeper& sleeper,
+      AsyncStreamFactory<ResponseType, RequestType>& stream_factory,
+      StreamReinitializer<ResponseType, RequestType>& reinitializer)
+      : retry_policy_prototype_(std::move(retry_policy)),
+        backoff_policy_prototype_(std::move(backoff_policy)),
+        sleeper_(std::move(sleeper)),
+        stream_factory_(std::move(stream_factory)),
+        reinitializer_(std::move(reinitializer)),
+        impl_(stream_factory_()) {}
+
+  ResumableAsyncStreamingReadWriteRpcImpl(ResumableStreamingReadRpc&&) = delete;
+  ResumableAsyncStreamingReadWriteRpcImpl& operator=(
+      ResumableStreamingReadRpc&&) = delete;
+
+  void Cancel() override {
+    // not guarding since this is the singular writer
```

Comment:
undo this. always guard member accesses and writes.

This is technically illegal and would probably be caught by tsan